File: Services/NetworkService.cs
using System;
using System.Diagnostics;
using System.Threading.Tasks;

namespace InvolveX.Cli.Services
{
    public class NetworkService
    {
        private readonly LogService _logService;

        public NetworkService(LogService logService)
        {
            _logService = logService;
        }

        public async Task<string> RunPingTest(string host)
        {
            try
            {
                // Validate host input
                if (string.IsNullOrWhiteSpace(host))
                {
                    return "Error: Host cannot be empty";
                }

                // Use ping with 4 packets and timeout for better reliability
                var arguments = $"-n 4 -w 5000 {host}";

                var process = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "ping",
                        Arguments = arguments,
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                process.Start();
                string stdout = await process.StandardOutput.ReadToEndAsync();
                string stderr = await process.StandardError.ReadToEndAsync();
                await process.WaitForExitAsync();

                _logService.Log($"[ping STDOUT]: {stdout}");
                _logService.Log($"[ping STDERR]: {stderr}");

                if (process.ExitCode != 0)
                {
                    return $"Ping test completed with warnings:\n{stdout}\n\nErrors:\n{stderr}";
                }

                return stdout;
            }
            catch (Exception ex)
            {
                _logService.Log($"Error running ping test: {ex.Message}");
                return $"Error running ping test: {ex.Message}";
            }
        }

        public async Task<string> RunSpeedTest()
        {
            _logService.Log("Running internet speed test using speedtest CLI.");

            try
            {
                // Check if speedtest CLI is installed
                var checkProcess = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "speedtest",
                        Arguments = "--version",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                checkProcess.Start();
                await checkProcess.WaitForExitAsync();

                if (checkProcess.ExitCode != 0)
                {
                    // Try to install speedtest CLI if not available
                    _logService.Log("Speedtest CLI not found, attempting to install...");

                    var installProcess = new Process
                    {
                        StartInfo = new ProcessStartInfo
                        {
                            FileName = "powershell",
                            Arguments = "winget install --id Ookla.Speedtest.CLI --accept-source-agreements --accept-package-agreements",
                            RedirectStandardOutput = true,
                            RedirectStandardError = true,
                            UseShellExecute = false,
                            CreateNoWindow = true
                        }
                    };

                    installProcess.Start();
                    string installOutput = await installProcess.StandardOutput.ReadToEndAsync();
                    string installError = await installProcess.StandardError.ReadToEndAsync();
                    await installProcess.WaitForExitAsync();

                    if (installProcess.ExitCode != 0)
                    {
                        return $"Failed to install speedtest CLI. Error: {installError}";
                    }

                    _logService.Log("Speedtest CLI installed successfully.");
                }

                // Run the speed test
                var speedTestProcess = new Process
                {
                    StartInfo = new ProcessStartInfo
                    {
                        FileName = "speedtest",
                        Arguments = "--accept-license --accept-gdpr",
                        RedirectStandardOutput = true,
                        RedirectStandardError = true,
                        UseShellExecute = false,
                        CreateNoWindow = true
                    }
                };

                speedTestProcess.Start();
                string output = await speedTestProcess.StandardOutput.ReadToEndAsync();
                string error = await speedTestProcess.StandardError.ReadToEndAsync();
                await speedTestProcess.WaitForExitAsync();

                if (speedTestProcess.ExitCode == 0)
                {
                    _logService.Log("Speed test completed successfully.");
                    return output;
                }
                else
                {
                    _logService.Log($"Speed test failed with exit code {speedTestProcess.ExitCode}: {error}");
                    return $"Speed test failed: {error}";
                }
            }
            catch (Exception ex)
            {
                _logService.Log($"Exception running speed test: {ex.Message}");
                return $"Error running speed test: {ex.Message}";
            }
        }
    }
}

File: services/NetworkService.js
const { spawn } = require('child_process');

class NetworkService {
  constructor(logService) {
    this.logService = logService;
  }

  async runPingTest(host, timeoutMs = 5000) {
    try {
      // Validate host input
      if (!host || typeof host !== 'string') {
        return 'Error: Host cannot be empty';
      }

      // Use ping command with timeout
      const args =
        process.platform === 'win32'
          ? ['-n', '4', '-w', Math.floor(timeoutMs / 1000).toString(), host]
          : ['-c', '4', '-W', Math.floor(timeoutMs / 1000).toString(), host];

      const result = await this.runProcess('ping', args);

      if (result.code === 0) {
        return result.stdout;
      } else {
        return `Ping test completed with warnings:\n${result.stdout}\n\nErrors:\n${result.stderr}`;
      }
    } catch (error) {
      this.logService.log(`Error running ping test: ${error.message}`);
      return `Error running ping test: ${error.message}`;
    }
  }

  async runSpeedTest() {
    this.logService.log('Running internet speed test using speedtest-cli.');

    try {
      const result = await this.runProcess('speedtest', ['--json']);

      if (result.code === 0) {
        const speedtestOutput = JSON.parse(result.stdout);

        if (
          speedtestOutput &&
          speedtestOutput.ping &&
          speedtestOutput.download &&
          speedtestOutput.upload
        ) {
          const formattedResult =
            `Ping: ${speedtestOutput.ping.latency.toFixed(2)} ms (Jitter: ${speedtestOutput.ping.jitter.toFixed(2)} ms)\n` +
            `Download: ${(speedtestOutput.download.bandwidth / 125000).toFixed(2)} Mbit/s\n` + // Convert bytes/sec to Mbit/s
            `Upload: ${(speedtestOutput.upload.bandwidth / 125000).toFixed(2)} Mbit/s\n\n` + // Convert bytes/sec to Mbit/s
            `Server: ${speedtestOutput.server.name} (${speedtestOutput.server.location}, ${speedtestOutput.server.country})\n` +
            `ISP: ${speedtestOutput.isp}`;

          this.logService.log('Speed test completed successfully.');
          return formattedResult;
        } else {
          this.logService.log(`Speed test output was not in expected format: ${result.stdout}`);
          return 'Error: Speed test output was not in expected format.';
        }
      } else {
        this.logService.log(`Speed test command failed with code ${result.code}: ${result.stderr}`);
        return `Error running speed test: ${result.stderr || 'Unknown error'}`;
      }
    } catch (error) {
      this.logService.log(`Exception running speed test: ${error.message}`);
      return `Error running speed test: ${error.message}`;
    }
  }

  async runProcess(command, args) {
    return new Promise((resolve, reject) => {
      const process = spawn(command, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: true,
      });

      let stdout = '';
      let stderr = '';

      process.stdout.on('data', data => {
        stdout += data.toString();
      });

      process.stderr.on('data', data => {
        stderr += data.toString();
      });

      process.on('close', code => {
        this.logService.log(`[${command} STDOUT]: ${stdout}`);
        this.logService.log(`[${command} STDERR]: ${stderr}`);

        resolve({
          code,
          stdout,
          stderr,
        });
      });

      process.on('error', error => {
        reject(error);
      });
    });
  }
}

module.exports = NetworkService;
